File: in_clojure/day_02_part_2.clj
(require '[clojure.string :as str])

(def input
  (-> "../inputs/day02.txt"
      (slurp)
      (str/trim)
      (str/split #",")
      (->> (mapv read-string))))

(def target 19690720)

(def ops {1 +, 2 *, 99 :return})

(defn execute [program-init breaking-point]
  (loop [pos 0 program program-init]
    (let [op-id (get program pos)
          op (get ops op-id)]
      (if (or (= op :return)
              (>= (first program) breaking-point))
        program
        (let [a*   (get program (+ 1 pos))
              b*   (get program (+ 2 pos))
              res* (get program (+ 3 pos))
              res  (op (get program a*)
                       (get program b*))]
          (recur (+ pos 4) (assoc program res* res)))))))

(def result
  (first
   (for [noun (range 100)
         verb (range 100)
         :let [tape (execute (assoc input 1 noun 2 verb) target)]
         :when (= target (first tape))]
     {:noun noun
      :verb verb})))

(println "Day 02 - Part 2:" (+ (* 100 (:noun result))
                               (:verb result)))

File: in_ruby/day_02_part_2.rb
input = File.read("../inputs/day02.txt").split(",").map(&:to_i)

TARGET=19690720

ADD=1
MULT=2
RETURN=99

def execute!(prog,limit)
  prog = prog.clone
  pos = 0
  loop do
    op_id,a_star,b_star,res_star = prog[pos .. pos+3]
    break if op_id == RETURN
    break if prog[0] > limit
    op = case op_id
         when ADD  then :+.to_proc
         when MULT then :*.to_proc
         end
    prog[res_star] = op[prog[a_star], prog[b_star]]
    pos+=4
  end
  prog
end

def find_noun_and_verb(input, target)
  99.times do |noun|
    99.times do |verb|
      prog = input.clone
      prog[1] = noun
      prog[2] = verb
      result = execute!(prog, target)
      if result[0] == target
        return [noun, verb]
      end
    end
  end
end

noun, verb = find_noun_and_verb(input, TARGET)

puts "Day 02 - Part 2: #{100*noun + verb}"
